ingestion-beam/src/main/java/com/mozilla/telemetry/decoder/ParseUri.java

```diff
@@ -54,13 +72,166 @@ private ParseUri() {
+        putBool("manual_download"), putBool("had_old_install"), putString("old_version"),
+        putString("old_build_id"), putString("version"), putString("build_id"),
+        putBool("default_path"), putBool("admin_user"),
+        // Default browser status code
+        putBoolPerCode(ImmutableMap.of("new_default", 1, "old_default", 2)),
+        // Default browser setting code
+        putBoolPerCode(ImmutableMap.of("set_default", 2)), //
+        putString("download_ip"), putString("attribution"),
+        putIntegerAsString("profile_cleanup_prompt"), putBool("profile_cleanup_requested"));
+
+    private static BiConsumer<String, JSONObject> ignore() {
+      return (value, payload) -> {
+      };
+    }
+
+    private static BiConsumer<String, JSONObject> putString(String key) {
+      return (value, payload) -> payload.put(key, value);
+    }
+
+    private static BiConsumer<String, JSONObject> appendString(String key, String separator) {
+      return (value, payload) -> payload.put(key,
+          Optional.ofNullable(payload.optString(key, null)).map(v -> v + separator).orElse("")
+              + value);
+    }
+
+    private static BiConsumer<String, JSONObject> putBool(String key) {
+      return (value, payload) -> payload.put(key, value.equals("1"));
+    }
+
+    private static BiConsumer<String, JSONObject> putInteger(String key) {
+      return (value, payload) -> payload.put(key, Integer.parseInt(value));
+    }
+
+    private static BiConsumer<String, JSONObject> putIntegerAsString(String key) {
+      return (value, payload) -> payload.put(key, Integer.toString(Integer.parseInt(value)));
+    }
+
+    private static BiConsumer<String, JSONObject> putBoolPerCode(Map<String, Integer> fieldCodes) {
+      return (string, payload) -> {
+        Integer value = Integer.parseInt(string);
+        fieldCodes.forEach((key, code) -> payload.put(key, value.equals(code)));
+      };
+    }
+
+    private static BiConsumer<String, JSONObject> putBoolPerCodeSet(
+        Map<String, Set<Integer>> fieldCodes) {
+      return (string, payload) -> {
+        Integer value = Integer.parseInt(string);
+        fieldCodes.forEach((key, set) -> payload.put(key, set.contains(value)));
+      };
+    }
+
+    private static JSONObject parsePingVersion(String[] elements) throws InvalidUriException {
+      // Parse ping version using a regex pattern
+      String pingVersion = elements[0];
+      Matcher matcher = PING_VERSION_PATTERN.matcher(pingVersion);
+      if (!matcher.find()) {
+        throw new UnknownPingVersionException(pingVersion);
+      }
+      // Check length of elements is valid for this version
+      int unexpectedElements = elements.length - SUFFIX_LENGTH.get(matcher.group(VERSION));
+      if (unexpectedElements != 0) {
+        throw new UnexpectedPathElementsException(unexpectedElements);
+      }
+      // Initialize new payload with ping version and funnelcake
```

Comment:
Is there documentation we can link here for what funnelcake is?

Reply:
hmm, seems like https://wiki.mozilla.org/Funnelcake should do
```suggestion
      // Initialize new payload with ping version and funnelcake ID
      // https://wiki.mozilla.org/Funnelcake
```